libs/librepcb/projecteditor/boardeditor/fsm/bes_addvia.cpp

```diff
@@ -283,45 +292,65 @@ BES_Base::ProcRetVal BES_AddVia::processSceneEvent(BEE_Base* event) noexcept {
 }
 
 bool BES_AddVia::addVia(Board& board) noexcept {
-  Q_ASSERT(mUndoCmdActive == false);
-  Q_ASSERT(mCurrentViaNetSignal);
+  Q_ASSERT(mSubState == SubState_Idle);
+  Q_ASSERT(!mUndoStack.isCommandGroupActive());
 
   try {
+    mSubState = SubState_PositioningVia;
     mUndoStack.beginCmdGroup(tr("Add via to board"));
-    mUndoCmdActive = true;
-    CmdBoardNetSegmentAdd* cmdAddSeg =
-        new CmdBoardNetSegmentAdd(board, *mCurrentViaNetSignal);
+    CmdBoardNetSegmentAdd* cmdAddSeg = nullptr;
+    if (mCurrentViaNetSignal) {
+      cmdAddSeg = new CmdBoardNetSegmentAdd(board, *mCurrentViaNetSignal);
+    } else {
+      NetSignal* closestSignal = getClosestNetSignal(board, Point(0, 0));
+      if (!closestSignal) {
+        abortPlacement();
+        return false;
+      }
+      cmdAddSeg = new CmdBoardNetSegmentAdd(board, *closestSignal);
+    }
     mUndoStack.appendToCmdGroup(cmdAddSeg);
     BI_NetSegment* netsegment = cmdAddSeg->getNetSegment();
     Q_ASSERT(netsegment);
-    CmdBoardNetSegmentAddElements* cmdAddVia =
-        new CmdBoardNetSegmentAddElements(*netsegment);
+    QScopedPointer<CmdBoardNetSegmentAddElements> cmdAddVia(
+        new CmdBoardNetSegmentAddElements(*netsegment));
     mCurrentVia = cmdAddVia->addVia(Point(0, 0), mCurrentViaShape,
                                     mCurrentViaSize, mCurrentViaDrillDiameter);
     Q_ASSERT(mCurrentVia);
-    mUndoStack.appendToCmdGroup(cmdAddVia);
+    mUndoStack.appendToCmdGroup(cmdAddVia.take());
     mViaEditCmd.reset(new CmdBoardViaEdit(*mCurrentVia));
     return true;
   } catch (Exception& e) {
-    if (mUndoCmdActive) {
-      try {
-        mUndoStack.abortCmdGroup();
-      } catch (...) {
-      }
-      mUndoCmdActive = false;
-    }
+    abortPlacement();
     QMessageBox::critical(&mEditor, tr("Error"), e.getMsg());
     return false;
   }
 }
 
+void BES_AddVia::abortPlacement(const bool showErrorMessage) noexcept {
+  if (mSubState != SubState_Idle) {
+    Q_ASSERT(mUndoStack.isCommandGroupActive());
+    try {
+      mUndoStack.abortCmdGroup();
+    } catch (Exception& e) {
+      if (showErrorMessage) {
+        QMessageBox::critical(&mEditor, tr("Error"), e.getMsg());
+      }
+    }
+  }
+  Q_ASSERT(!mUndoStack.isCommandGroupActive());
```

Comment:
Same here - if we are not in `SubState_PositioningVia`, we must not rely on whether no command is active currently.

```suggestion
```